roles/tests-integration/tests/common/sniffer.rs

```diff
@@ -0,0 +1,403 @@
+        if let Ok((receiver_from_client, send_to_client, _, _)) =
+            Connection::new::<'static, AnyMessage<'static>>(
+                stream,
+                HandshakeRole::Initiator(initiator),
+            )
+            .await
+        {
+            Some((receiver_from_client, send_to_client))
+        } else {
+            None
+        }
+    }
+
+    async fn recv_from_down_send_to_up(
+        recv: Receiver<MessageFrame>,
+        send: Sender<MessageFrame>,
+        downstream_messages: MessagesAggregator,
+    ) -> Result<(), SnifferError> {
+        while let Ok(mut frame) = recv.recv().await {
+            let (msg_type, msg) = Self::message_from_frame(&mut frame);
+            downstream_messages.add_message(msg_type, msg);
+            if send.send(frame).await.is_err() {
+                return Err(SnifferError::UpstreamClosed);
+            };
+        }
+        Err(SnifferError::DownstreamClosed)
+    }
+
+    async fn recv_from_up_send_to_down(
+        recv: Receiver<MessageFrame>,
+        send: Sender<MessageFrame>,
+        upstream_messages: MessagesAggregator,
+    ) -> Result<(), SnifferError> {
+        while let Ok(mut frame) = recv.recv().await {
+            let (msg_type, msg) = Self::message_from_frame(&mut frame);
+            upstream_messages.add_message(msg_type, msg);
+            if send.send(frame).await.is_err() {
+                return Err(SnifferError::DownstreamClosed);
+            };
+        }
+        Err(SnifferError::UpstreamClosed)
+    }
+
+    fn message_from_frame(frame: &mut MessageFrame) -> (MsgType, AnyMessage<'static>) {
+        match frame {
+            Frame::Sv2(frame) => {
+                if let Some(header) = frame.get_header() {
+                    let message_type = header.msg_type();
+                    let mut payload = frame.payload().to_vec();
+                    let message: Result<AnyMessage<'_>, _> =
+                        (message_type, payload.as_mut_slice()).try_into();
+                    match message {
+                        Ok(message) => {
+                            let message = Self::into_static(message);
+                            (message_type, message)
+                        }
+                        _ => {
+                            println!(
+                                "Received frame with invalid payload or message type: {frame:?}"
+                            );
+                            panic!();
+                        }
+                    }
+                } else {
+                    println!("Received frame with invalid header: {frame:?}");
+                    panic!();
+                }
+            }
+            Frame::HandShake(f) => {
+                println!("Received unexpected handshake frame: {f:?}");
+                panic!();
+            }
+        }
+    }
+
+    fn into_static(m: AnyMessage<'_>) -> AnyMessage<'static> {
+        match m {
```

Comment:
Can use this: 
```
macro_rules! convert_message {
    ($enum:ident, $variant:ident, { $($sub_variant:ident),* }) => {
        match $variant {
            $(
                $enum::$sub_variant(m) => $enum::$sub_variant(m.into_static()),
            )*
        }
    };
    ($enum:ident, { $($variant:ident),* }) => {
        match $enum {
            $(
                $enum::$variant(m) => $enum::$variant(m.into_static()),
            )*
        }
    };
}

```
This might reduce the bulk